src/protocol/client.c

```diff
@@ -153,247 +391,234 @@ protocolClientNew(const String *name, const String *service, IoRead *read, IoWri
     }
-    MEM_CONTEXT_TEMP_END();
 
-    FUNCTION_LOG_RETURN_VOID();
+    FUNCTION_LOG_RETURN(PACK_READ, protocolClientResponseInternal(this));
 }
 
 /**********************************************************************************************************************************/
-FN_EXTERN void
-protocolClientCommandPut(ProtocolClient *const this, ProtocolCommand *const command, const bool dataPut)
+FN_EXTERN PackRead *
+protocolClientSessionRequest(ProtocolClientSession *const this, const ProtocolClientRequestParam param)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
-        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT, this);
-        FUNCTION_LOG_PARAM(PROTOCOL_COMMAND, command);
-        FUNCTION_LOG_PARAM(BOOL, dataPut);
+        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT_SESSION, this);
+        FUNCTION_LOG_PARAM(PACK_WRITE, param.param);
     FUNCTION_LOG_END();
 
     ASSERT(this != NULL);
-    ASSERT(command != NULL);
 
-    // Expect idle state before command put
-    protocolClientStateExpect(this, protocolClientStateIdle);
+    protocolClientRequestInternal(this, protocolCommandTypeProcess, param.param);
 
-    // Switch state to cmd-put
-    this->state = protocolClientStateDataPut;
+    FUNCTION_LOG_RETURN(PACK_READ, protocolClientResponseInternal(this));
+}
 
-    // Put command
-    protocolCommandPut(command, this->write);
+/**********************************************************************************************************************************/
+FN_EXTERN void
+protocolClientSessionRequestAsync(ProtocolClientSession *const this, const ProtocolClientRequestParam param)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT_SESSION, this);
+        FUNCTION_LOG_PARAM(PACK_WRITE, param.param);
+    FUNCTION_LOG_END();
 
-    // Switch state to data-get/data-put after successful command put
-    this->state = dataPut ? protocolClientStateCommandDataGet : protocolClientStateDataGet;
+    ASSERT(this != NULL);
+    ASSERT(this->async);
 
-    // Reset the keep alive time
-    this->keepAliveTime = timeMSec();
+    protocolClientRequestInternal(this, protocolCommandTypeProcess, param.param);
 
     FUNCTION_LOG_RETURN_VOID();
 }
 
 /**********************************************************************************************************************************/
 FN_EXTERN PackRead *
-protocolClientExecute(ProtocolClient *const this, ProtocolCommand *const command, const bool resultRequired)
+protocolClientSessionResponse(ProtocolClientSession *const this)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
-        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT, this);
-        FUNCTION_LOG_PARAM(PROTOCOL_COMMAND, command);
-        FUNCTION_LOG_PARAM(BOOL, resultRequired);
+        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT_SESSION, this);
     FUNCTION_LOG_END();
 
     ASSERT(this != NULL);
-    ASSERT(command != NULL);
+    ASSERT(this->async);
```

Comment:
> protocolClientSession is an inherently async API, and non-async consumers would be using protocolClientResponse[P]? instead?

Most callers just use protocolClientRequestP(). This returns data in a synchronous fashion without needing a session. protocolClientSessionRequest() returns data synchronously when a session is needed (Db is the only caller here). protocolClientSessionRequestAsync()/protocolClientSessionResponse() is for async requests (all of them require a session or they couldn't be async).

> it confuses me a little why there's not a corresponding protocolClientSessionResponseAsync, say.

Because the response is not async -- it blocks. So we can send an async request and do something else but once we need the response we wait until it is available. Under the covers, protocolClientSessionRequest() is just doing an async request + blocking response, but the idea is to keep the interface simple for the simpler cases. Most callers don't need to worry about async (or sessions for that matter).

In theory every sync request could open a session, but this is less efficient in the protocol so we have different interfaces for performance.

Reply:
Makes sense, thanks.